Connection: move the connection owner deletion notice to the Connection package

kbrown9 opened the pull request:
A notice displays when a user attempts to delete the connection owner. Move this notice and the endpoint for setting the connection owner to the connection package. With this change, the connection owner deletion notice will be displayed for all consumers of the connection package.

Fixes #15568

#### Changes proposed in this Pull Request:
* Move the `delete_user_update_connection_owner_notice` method from the JITM package to the Connection package.
* Move the `/connection/owner` endpoint, including the callback and permission callback, to the Connection package.

#### Jetpack product discussion
* n/a

#### Does this pull request change what data or activity we track or use?
* no

#### Testing instructions:
##### Attempt to delete the connection owner with a logged-out admin.
1. Install, activate, and connect Jetpack. This user is the primary admin and the current connection owner.
2. Checkout this PR's branch.
3. Navigate to `Users -> Add New` and create another administrator user. This user is the secondary admin. Don't connect this user yet.
4. While logged in as the secondary admin, navigate to the Users page and attempt to delete the primary user (connection owner). You should see a notice that warns you that you're attempting to delete the connection owner.
5. Click the `Connect to WordPress.com` button to connect the secondary admin to WPCOM.
6. Refresh the user deletion page from step 4 and click the `Set new connection owner` button to change the connection owner. (Don't delete the primary admin user, because we'll switch the connection owner back to that user in the next test.)
7. The secondary admin is now the connection owner. You can verify this by navigating to `Tools -> Site Health`, selecting the `Info` tab, and opening the Jetpack drop-down. The `Jetpack Master User` (aka connection owner) should be the secondary admin.

##### Attempt to delete the connection owner with a logged-in admin.
1. Log in as the primary admin from step 1 above. This user should already be connected to WordPress.com
2. Navigate to Users and attempt to delete the connection owner, the secondary admin.
3. Since the primary admin is already connected to WordPress.com, you should see the `Set new connection owner` button in the notice. Click that button.
4. The primary admin is now the connection owner. You can verify this by navigating to `Tools -> Site Health`, selecting the `Info` tab, and opening the Jetpack drop-down. The `Jetpack Master User` (aka connection owner) should be the primary admin.

After conducting these tests, verify that no errors were generated in the debug.log. You can also verify that the `jetpack_connection_owner_notice_view` and `jetpack_set_connection_owner_success` tracks events were recorded.

##### Verify that the `connection/owner` endpoint checks user permissions properly.
1. Send a POST request to the `jetpack/v4/connection/owner` endpoint while not logged in.
2. Verify that a 401 response is received with the code `invalid_user_permission_set_connection_owner`.

#### Proposed changelog entry for your changes:
* tbd

kbrown9 committed:
Connection: move the connection owner deletion notice to the Connection package

A notice displays when a user attempts to delete the connection owner. Move this notice
and the endpoint for setting the connection owner to the connection package. With this change,
the connection owner deletion notice will be displayed for all consumers of the connection package.

jeherve reviewed (dismissed):
This tested well for me. 🚢

anomiex commented on tests/php/_inc/lib/test_class.rest-api-endpoints.php:
```diff
@@ -960,40 +960,78 @@ public function test_fetch_not_registered_widget_data() {
 
 	/**
-	 * Test changing the master user.
+	 * Test changing the master user without the owner parameter.
```
Do these tests still belong here, or should they be moved along with the implementation of the endpoint they're testing?

kbrown9 replied:
I don't think it's currently possible to implement these tests in the Connection package tests. These tests require creating users and saving them in the database. The Connection package tests do not run in a full WordPress environment with a database. Instead, they use the [WorDBless package](https://github.com/Automattic/wordbless), which provides WordPress core functions without a database.

So, I left the tests the require a WordPress environment in the Jetpack plugin's tests and added a few tests that do not require the  WordPress environment to the Connection package.

anomiex reviewed (commented):
Asking a few questions to try to understand this better.

kbrown9 committed:
Connection: minor changes to connection owner deletion notice

- Remove the `method_exists` check for Tracking:::record_user_event.
- Add error handling for the Ajax request.
- Use Tracking::class insstead of the full class name.
- Remove a duplicate unit test and update the unit test comments.

kbrown9 commented on packages/connection/src/class-owner.php:
```diff
@@ -0,0 +1,184 @@
+
+		// Track it!
+		$tracking = new Tracking();
+		if ( method_exists( $tracking, 'record_user_event' ) ) {
```
I checked the [PR](https://github.com/Automattic/jetpack/pull/13388) where this check was added, and it looks like the intention was just to be overly defensive. Since the `Automattic\Jetpack\Tracking` class has always had the `record_user_event` method, I removed this `method_exists` check in commit e6540b8.

kbrown9 commented on packages/connection/src/class-owner.php:
```diff
@@ -0,0 +1,184 @@
+							},
+						} ).done( function() {
+							submitBtn.disabled = false;
+						} );
```
In commit e6540b8, I added error handling to the Ajax call. If the call fails, an error message will be displayed and the button will be enabled.

kbrown9 commented on packages/connection/src/class-rest-connector.php:
```diff
@@ -278,4 +290,101 @@ public function connection_reconnect() {
+		if ( $updated && ! $xml->isError() ) {
+
+			// Track it.
+			if ( class_exists( 'Automattic\Jetpack\Tracking' ) ) {
```
I changed this in commit  e6540b8.

anomiex commented on packages/connection/src/class-owner.php:
```diff
@@ -0,0 +1,187 @@
+							submitBtn.disabled = false;
+						} ).fail( function() {
+							results.classList.add( 'error-message' );
+							results.innerHTML = "<?php esc_html_e( 'Something went wrong. Please try again.', 'jetpack' ); ?>";
```
Honest question: Is this the preferred behavior for error reporting? My inclination would be to try to relay a more specific message from the endpoint. But maybe that would be more confusing for a typical user?

jeherve commented:
I'll close this PR for now because of the lack of activity on this. We can always reopen in the future if needed, but it will need a rebase, so it may be easier to start a new PR at this point.
